src/app/(proper_react)/(redesign)/(authenticated)/admin/churn-subscribers/ChurnAdmin.tsx

```diff
@@ -0,0 +1,121 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+"use client";
+
+import { useEffect, useState } from "react";
+import { useSession } from "next-auth/react";
+import styles from "./ChurnAdmin.module.scss";
+import { SubscriberChurnRow } from "knex/types/tables";
+import { getAllChurns, upsertAllChurns } from "./actions";
+
+export const ChurnAdmin = () => {
+  const session = useSession();
+  const [subscriberData, setSubscriberData] = useState<
+    SubscriberChurnRow[] | null
+  >(null);
+
+  // Load data on component mount
+  useEffect(() => {
+    const loadSubscriberData = async () => {
+      const data = await getAllChurns();
+      setSubscriberData(data);
+    };
+
+    void loadSubscriberData();
+  }, []);
+
+  const handleFileUpload = (event: React.ChangeEvent<HTMLInputElement>) => {
+    const file = event.target.files?.[0];
+    if (!file) return;
+
+    const reader = new FileReader();
+    reader.onload = async (e) => {
+      const text = e.target?.result as string;
+      const rows = text.split("\n");
+
+      console.log({ rows: rows.length });
+      const parsedData: SubscriberChurnRow[] = rows
+        .slice(1)
+        .filter((row) => row.trim()) // Skip empty rows
+        .map((row) => {
+          const values = row.split(",");
+          return {
+            userid: values[0],
+            customer: values[1],
+            created: values[2],
+            nickname: values[3],
+            intervl: values[4],
+            plan_id: values[6],
+            product_id: values[7],
+            current_period_end: new Date(values[8]).toISOString(),
+          };
+        });
+
+      console.log({ parsedData });
+      // Update backend and refresh data
+      await upsertAllChurns(parsedData);
```

Comment:
I've got a slight worry that starting the upload the moment the user selects a file is also prone to attract errors from simple misclicks. To be safe, we could wrap it in a form, add a submit button, and only upload it once the form is submitted?

Reply:
added this